### PR TITLE
Allows to pass Koncorde filters as a search query

### PR DESCRIPTION
**Pull request by Aschen**
## What does this PR do ?

Allows to send execute `document:search` by passing Koncorde filters instead of an ES query.  

The `lang` argument allows to specify which language the query is written on. By default it's `elasticsearch` but `koncorde` can be specified.

The `regexp` clause is not supported because ES is not compatible with regexes :expressionless: 

The `bool` operator is not supported for now because we don't use it a lot IMHO and I'm lazy

### How should this be manually tested?

<details><summary>Create collection and documents</summary>

```bash
kourou collection:create nyc-open-data yellow-taxi '{
  mappings: {
    properties: {
      city: { type: "keyword" },
      name: { type: "keyword" },
      age: { type: "integer" }
    }
  }
}'

kourou document:mCreateOrReplace nyc-open-data yellow-taxi '{
  documents: [
    {
      _id: "melis",
      body: {
        city: "Istanbul",
        name: "Melis",
        age: 25,
      }
    },
    {
      _id: "john",
      body: {
        city: "Istanbul",
        age: 25,
      }
    },
    {
      _id: "aschen",
      body: {
        city: "Antalya",
        name: "Aschen",
        age: 27,
      }
    }
  ]
}'
```

</details>

Then execute this Koncorde query:

```bash
kourou sdk:execute '
  return sdk.document.query({
    controller: "document",
    action: "search",
    index: "nyc-open-data",
    collection: "yellow-taxi",

    lang: "koncorde",
    body: {
      query: {
        "and": [
          { "equals": { "city": "Istanbul" } },
          { "exists": "name" },
          {
            "not": {
              "range": { "age": { "gt": 28 } }
            }
          }
        ]
      }
    },
  })
'
```

### Other changes
 
  - [bugfix] `app` property from the `Controller` class is not private anymore

### CODE REVIEW

File: doc/2/api/controllers/document/search/index.md [hunk 1]
```diff
@@ -24,14 +24,23 @@ It can lead to memory leaks if a scroll duration too large is provided, or if to
 You can restrict the scroll session maximum duration under the `services.storage.maxScrollDuration` configuration key.
 :::
 
+<SinceBadge version="change-me"/>
+
+This method also support [Koncorde Filters DSL](/core/2/guides/cookbooks/realtime-api) to match documents by passing the `lang` argument with the value `koncorde`.
```
Comment (scottinet) [inline doc/2/api/controllers/document/search/index.md:29]:
```suggestion
This method also supports the [Koncorde Filters DSL](/core/2/guides/cookbooks/realtime-api) to match documents by passing the `lang` argument with the value `koncorde`.  
```

File: doc/2/api/controllers/document/search/index.md [hunk 2]
```diff
@@ -96,14 +106,15 @@ Method: GET
...
+- `lang`: specify the query language to use. By default, it's `elasticsearch` but `koncorde` can also be used
 
 ---
 
 ## Body properties
 
 ### Optional:
 
-- `query`: the search query itself, using the [ElasticSearch Query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl.html) syntax.
+- `query`: the search query itself, using the [ElasticSearch Query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl.html) or [Koncorde Filters DSL](/core/2/guides/cookbooks/realtime-api) syntax.
```
Comment (scottinet) [inline doc/2/api/controllers/document/search/index.md:117]:
```suggestion
- `query`: the search query itself, using the [ElasticSearch Query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl.html) or the [Koncorde Filters DSL](/core/2/guides/cookbooks/realtime-api) syntax.
```

File: lib/kerror/codes/2-api.json [hunk 3]
```diff
@@ -69,6 +69,18 @@
...
           "class": "BadRequestError"
+        },
+        "koncorde_unknown_keyword": {
+          "description": "An unknown keyword has been provided in filters",
+          "code": 12,
+          "message": "The keyword \"%s\" is not one of the Koncorde filters DSL keywords.",
+          "class": "BadRequestError"
+        },
+        "koncorde_restricted_keyword": {
+          "description": "An restricted keyword has been provided in filters",
```
Comment (scottinet) [inline lib/kerror/codes/2-api.json:80]:
```suggestion
          "description": "A restricted keyword has been provided in filters",
```

**Review by scottinet — CHANGES_REQUESTED** [review]
Shouldn't this new `lang` property also be made available to the following routes?
`document:deleteByQuery`
`document:updateByQuery`
`security:search*`

**Comment by Aschen** [comment]
@scottinet I added it to the other API actions that accepts a search query (`security:searchProfiles` and `security:searchRoles` don't)

**Review by scottinet — APPROVED** [review]

File: doc/2/api/controllers/security/search-api-keys/index.md [hunk 4]
```diff
@@ -12,14 +12,23 @@ Searches for a user API keys.
 To search for an API key corresponding to a token you can search on the `fingerprint` property which is a SHA256 hash of the token.
 :::
 
+<SinceBadge version="change-me"/>
```
Comment (xbill82) [inline doc/2/api/controllers/security/search-api-keys/index.md:15]: Are we tracking these things effectively so they don't get diluted in the soup?

File: lib/api/controller/base.js [hunk 5]
```diff
@@ -81,6 +81,48 @@ class NativeController extends BaseController {
...
+
+    try {
+      return await this.ask('core:storage:public:translate', koncordeFilters);
+    }
+    catch (error) {
+      throw kerror.get(
+        'api',
+        'assert',
+        'koncorde_restricted_keyword',
+        error.keyword.type,
```
Comment (xbill82) [inline lib/api/controller/base.js:106]: This is a weird catch... It's like the catch knows exactly the type of the error that's thrown from the `ask` call. It's some kind of hard coupling, IMHO. We should at least check the type of the error to ensure the subfield `keyword` exists inside it.

File: doc/2/api/controllers/security/search-api-keys/index.md (hunk 4 above)
Comment (Aschen) [reply doc/2/api/controllers/security/search-api-keys/index.md:15]: The release tool automatically update the badges with the released version

**Review by xbill82 — APPROVED** [review]

**Pull request merged**